pkg/helmreconciler/listeners.go

```diff
@@ -0,0 +1,472 @@
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// EndDelete delegates EndDelete to the Listeners in last to first order.
+func (l *CompositeRenderingListener) EndDelete(instance runtime.Object, err error) error {
+	// reverse order for completions
+	allErrors := make([]error, 0, len(l.Listeners))
+	for index := len(l.Listeners) - 1; index > -1; index-- {
+		if listenerErr := l.Listeners[index].EndDelete(instance, err); listenerErr != nil {
+			allErrors = append(allErrors, listenerErr)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// EndReconcile delegates EndReconcile to the Listeners in last to first order.
+func (l *CompositeRenderingListener) EndReconcile(instance runtime.Object, err error) error {
+	// reverse order for completions
+	allErrors := make([]error, 0, len(l.Listeners))
+	for index := len(l.Listeners) - 1; index > -1; index-- {
+		if listenerErr := l.Listeners[index].EndReconcile(instance, err); listenerErr != nil {
+			allErrors = append(allErrors, listenerErr)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// LoggingRenderingListener is a RenderingListener which logs events.  It also updates the HelmReconciler logger with
+// values applying to various stages of processing (e.g. chart=chart-name, kind=resource-kind, etc.).  This can be used
+// as the first listener in a CompositeRenderingListener.
+type LoggingRenderingListener struct {
+	reconciler  *HelmReconciler
+	Level       int
+	loggerStack []logr.Logger
+}
+
+var _ RenderingListener = &LoggingRenderingListener{}
+var _ ReconcilerListener = &LoggingRenderingListener{}
+
+// RegisterReconciler associates the HelmReconciler with the LoggingRenderingListener
+func (l *LoggingRenderingListener) RegisterReconciler(reconciler *HelmReconciler) {
+	l.reconciler = reconciler
+}
+
+// BeginReconcile logs the event
+func (l *LoggingRenderingListener) BeginReconcile(instance runtime.Object) error {
+	l.reconciler.logger.V(l.Level).Info("begin reconciling resources")
+	return nil
+}
+
+// BeginDelete logs the event
+func (l *LoggingRenderingListener) BeginDelete(instance runtime.Object) error {
+	l.reconciler.logger.V(l.Level).Info("begin deleting resources")
+	return nil
+}
+
+// BeginChart logs the event and updates the logger to log with values chart=chart-name
+func (l *LoggingRenderingListener) BeginChart(chart string, manifests []manifest.Manifest) ([]manifest.Manifest, error) {
+	l.loggerStack = append(l.loggerStack, l.reconciler.logger)
+	l.reconciler.logger = l.reconciler.logger.WithValues("chart", chart)
+	l.reconciler.logger.V(l.Level).Info("begin updating resources for chart")
+	return manifests, nil
+}
+
+// BeginResource logs the event and updates the logger to log with values resource=name, kind=kind, apiVersion=api-version
+func (l *LoggingRenderingListener) BeginResource(obj runtime.Object) (runtime.Object, error) {
+	l.loggerStack = append(l.loggerStack, l.reconciler.logger)
+	accessor := meta.NewAccessor()
+	kind, _ := accessor.Kind(obj)
```

Comment:
This is just setting up the logger for the call with value pairs.  If this errors, it should not result in an error for the install.  This is why the error is ignored.  It doesn't make sense to return an error if it doesn't result in an error that should be handled.  That is unless you think reconciliation should fail because we couldn't add metadata to the log entries?  That's a different discussion.

Reply:
Keep in mind that one of the goals is for other developers to be able to easily reason about the code. What you just explained above is not obvious from reading the code and leads to unnecessary head scratching for future contributors. If you have to spend a paragraph explaining why an error is ok to ignore, isn't it easier just to handle it?